setuptools/package_index.py

```diff
@@ -588,24 +588,27 @@ def find(req, env=None):
                     continue
 
                 if dist in req and (dist.precedence<=SOURCE_DIST or not source):
-                    return dist
+                    mylocation = self.download(dist.location, tmpdir)
+                    if os.path.exists(mylocation):
+                        return dist, mylocation
+            return None, None
```

Comment:
The presence of `None, None` is a bad smell to me. Better would be to return above a single value which wraps the necessary components, maybe a `namedtuple('DistLocation', 'dist location')`.